opds_import.py

```diff
@@ -929,7 +952,18 @@ def extract_feed_data(self, feed, feed_url=None):
         metadata = {}
         circulationdata = {}
         for id, m_data_dict in fp_metadata.items():
-            external_identifier, ignore = Identifier.parse_urn(self._db, id)
+            xml_data_dict = xml_data_meta.get(id, {})
+
+            external_identifier = None
+            if custom_identifier == ExternalIntegration.DCTERMS_IDENTIFIER:
+                dcterms_ids = xml_data_dict.get('dcterms_identifiers', [])
+                if len(dcterms_ids) > 0:
+                    external_identifier, ignore = Identifier.for_foreign_id(
+                            self._db, dcterms_ids[0].type, dcterms_ids[0].identifier)
+
+            if external_identifier is None:
+                external_identifier, ignore = Identifier.parse_urn(self._db, id)
+
```

Comment:
If there are multiple <dcterms:identifier> tags, I would expect _all_ of them to be picked up and converted into IdentifierData objects. That's because we may need those identifiers later on, for things like finding book covers, or determining that two different books are the same literary work.

The metadata layer will treat the _first_ IdentifierData in a list as the primary identifier--the one you're creating with this code. I would create a helper method that handles the `identifier_mapping`, `identified_failures`, `identifier_obj` stuff, and then call it for every item in the the `dcterms_identiifer` list, regardless of the value of `custom_identiifer`.

Then, if `custom_identifier` is _not_ set to `DCTERMS_IDENTIFIER`, I would parse the `<id>`, run it through the helper method, and insert the resulting `IdentifierData` at the _beginning_ of the list, making it the primary.

Basically, everything in `dcterms_identifier` should become an `IdentifierData`, with the `<id>`  _maybe_ becoming an `IdentifierData` as well.

Reply:
Looking more closely at the code , I see that that `dcterms_identifiers` already contains a bunch of `IdentifierData` objects. Those objects will be passed in to the `Metadata` constructor as the `identifiers` argument, so the imported book will in fact end up with a number of associated `Identifier`s. So that part is already fine; I was wrong about what this code is doing.

All you need to do is change what ends up as 'primary_identiifer'. Right now you do that by replacing `external_identifier`. That's a little awkward because an `IdentifierData` is becoming an `Identifier` and then immediately becoming an `IdentifierData` again. But it works fine.

So this could use a comment to explain that you're seeing which of the IdentifierData objects should become the primary identifier, but other than that no code changes are necessary.

Reply:
Once we just can take the `dcterms:identifier` from the metadata (limitations on the feedparse library), I couldn't use the value before transform in a `IdentifierData`.
I'm commenting this trying to make clear.

Also, I discovery that <id> was being ignored, i'm fixing it too.